### PR TITLE
Testing guide page

### LINKED ISSUE #3: Design Page

### LINKED ISSUE #4: Team Process page

### LINKED ISSUE #7: Testing Guide

### PR DESCRIPTION
**Pull request by Ullaakut**
# Goal of this PR

Fixes #7 
Fixes #4 
Fixes #3 

This PR also adds an MIT license to the repository, and a simple description of what this repo is about in the README, as well as a number of features to our mkdocs setup:

* Admonitions, which enable us to write notes, examples, warnings etc. that can be folded or expanded.
* Markdown includes, which keep our documentation tidy by including external files into our compiled markdown before it is rendered into a website.
* MkDocs excludes, which prevents MkDocs from trying to compile the files we want to include.
* Automatically detects whether machine uses dark or light theme and adjusts the palette of the docs to it. Also adds a button at the top right of the site to manually toggle between dark/light.

## Examples

This is a warning admonition that is expanded by default.

SharedScreenshot

These are example admonitions which contain code examples and are collapsed by default. Users can expand them to read the code, and fold them back.

SharedScreenshot

## Additional Notes

I will explain how to use admonitions and imports for this repository when implementing the section about documentation guidelines in issue #6 

## How to test it

Dependencies: Python3, Pip.

* Check out this branch on your machine
* `pip install -r requirements.txt`
* `mkdocs build`
* Open `site/testing/index.html` in whatever browser.

### CODE REVIEW

**Comment by ricardomgoncalves** [comment]
Looks very good.

File: content/include/generic.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+            binary.BigEndian.PutUint64(address[0:], random.Uint64())
+    
+            addresses = append(addresses, address)
+        }
+    
+        return addresses
+    }
+    
+    func GenericAddress(index int) flow.Address {
+    return GenericAddresses(index + 1)[index]
```
[the remark this answers is not in the document]
Comment (Ullaakut) [reply content/include/generic.md:17]: Indentation issue indeed, it seems copy pasting broke it! Nice catch :)

**Comment by Ullaakut** [comment]
@ricardomgoncalves Thanks for the review! It should be all fixed now. I also removed trailing whitespaces in the snippets. LMK if it looks good to you now :)

**Comment by Ullaakut** [comment]
I added the contents of #4 to this PR since they are quite small and I don't want to deal with the latency of waiting for extra reviews for such a small amount of changes.

**Comment by Ullaakut** [comment]
Added contents of #3 for similar reasons.

File: content/testing.md [hunk 2]
```diff
@@ -1,23 +1,158 @@
 # Testing Guide
 
-TODO: Explain why it is essential to have proper testing in our projects and for PRs to maintain tests up-to-date.
+At Optakt, we consistently write tests to ensure a reliable engineering environment where quality is paramount. Over the course of the product development life cycle, testing saves time and money, and helps developers write better code, more efficiently. Untested code is fragile, difficult to maintain and becomes questionable as soon as changes are made.
```
Comment (Maelkum) [inline content/testing.md:3]: General question - does documentation generation work with a 'single sentence per line' style?

File: content/testing.md [hunk 3]
```diff
@@ -1,23 +1,158 @@
...
+
+### Parallelization
+
+Since the version `1.7` of Go, tests can be run in parallel. This can be done by calling [`t.Parallel`](https://pkg.go.dev/testing#T.Parallel) in each subtest. Calling this function signals that the test is to be run in parallel with (and only with) other parallel tests, and the amount of tests running in parallel is limited by the value of [`runtime.GOMAXPROCS`](https://pkg.go.dev/runtime#GOMAXPROCS).
+
+There are multiple advantages to parallelizing tests:
+
+* It ensures that regardless of the order in which inputs are given, components behave as expected.
+* It maximizes performance, which in turns results in a faster CI and a faster workflow for everyone.
+* By making tests run faster, it allows you to write more tests and therefore produce more actionable data to find bugs as well as improve tests cases and coverage.
```
Comment (Maelkum) [inline content/testing.md:68]: I don't think this bullet point adds much value tbh, it kind of falls in the "faster workflow" category of benefits.

File: content/include/table_driven_parallel.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+    func TestParseCadenceArgument(t *testing.T) {
+        tests := []struct {
```
Comment (Maelkum) [inline content/include/table_driven_parallel.md:2]: I assume it's by intent to use a real-world example of a function.. But I wonder if having this large test setup detracts from the point being made about the loop variable.. The point of the whole thing is the single line no. 28.

File: content/testing.md [hunk 5]
```diff
@@ -1,23 +1,158 @@
...
+* It makes it possible to ensure that the components you test are concurrency-safe
+
+!!! danger "Parallelizing table-driven tests"
+
+    When it comes to table-driven tests, a common caveat developers fall into is to call `t.Parallel` in their subtest without capturing the loop variable with their test case.
```
Comment (Maelkum) [inline content/testing.md:73]: Can you fall into a caveat? You fall/run into into a *pitfall*, you need to be aware of the *caveat*?
I may be wrong, genuinely not sure :D

File: content/testing.md [hunk 6]
```diff
@@ -1,23 +1,158 @@
...
+
+??? example "index_internal_test.go"
+
+    ```go
+{!index_internal_test.md!}
+    ```
+
+#### Table-Driven Tests
+
+It makes a lot of sense to use subtests when testing the behavior of complex components, but it is better to use [table-driven tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) when testing a simple function with a given output for a given input, and that there are many cases to cover. For such cases, table-driven tests massively improve clarity and readability.
```
Comment (Maelkum) [inline content/testing.md:97]:
```suggestion
It makes a lot of sense to use subtests when testing the behavior of complex components, but it is better to use [table-driven tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) when testing a simple function with a given output for a given input, and there are many cases to cover. For such cases, table-driven tests massively improve clarity and readability.
```

Also, dubious about "testing a simple function with a given output for a given input". Perhaps it's better to have it as an "expected" output with a given input.

File: content/testing.md [hunk 7]
```diff
@@ -1,23 +1,158 @@
...
+```go
+//go:build integration
+// +build integration
+
+package dps_test
+```
+
+## Examples
+
+In Go, good package documentation includes not only comments for each public type and method, but also runnable examples and benchmarks in some cases. Godoc allows defining  examples which are verified by running them as tests and can be manually launched by readers of the documentation on the package's Godoc webpage.
```
Comment (Maelkum) [inline content/testing.md:150]:
```suggestion
In Go, good package documentation includes not only comments for each public type and method, but also runnable examples and benchmarks in some cases. Godoc allows defining examples which are verified by running them as tests and can be manually launched by readers of the documentation on the package's Godoc webpage.
```

(two consecutive spaces before "examples which are verified...")

**Review by Maelkum — APPROVED** [review]
This looks awesome, great work! 😃

File: content/testing.md (hunk 2 above)
Comment (Ullaakut) [reply content/testing.md:3]: Yes it does. I forgot about using that here, but it's definitely something I should add to the docs style guide and do throughout these docs! Nice catch :)

File: content/testing.md (hunk 3 above)
Comment (Ullaakut) [reply content/testing.md:68]: There is the advantage of random ordering and concurrent execution upon the same component though. I agree that two points are about the speed, but the two others are about completely different subjects, are they not? 😄

**Comment by Maelkum** [comment]
One thing that perhaps goes without saying, but then again, perhaps it doesn't.. is minding the code coverage. If there's new code that adds some new functionality and decreases the percentage of code covered by tests, there should be accompanying unit tests for it. Ideally a PR should not decrease code coverage of the project/package, only increase it.

File: content/include/table_driven_parallel.md (hunk 4 above)
Comment (Ullaakut) [reply content/include/table_driven_parallel.md:2]: True that. I'll change it and use the example from the testing package itself, they have a nice concise example there :)

File: content/testing.md (hunk 5 above)
Comment (Ullaakut) [reply content/testing.md:73]: You're right, pitfall is the word I should have used hehe.

File: content/testing.md (hunk 7 above)
Comment (Ullaakut) [reply content/testing.md:150]: Nice catch man!! :D

**Comment by Ullaakut** [comment]
> One thing that perhaps goes without saying, but then again, perhaps it doesn't.. is minding the code coverage. If there's new code that adds some new functionality and decreases the percentage of code covered by tests, there should be accompanying unit tests for it. Ideally a PR should not decrease code coverage of the project/package, only increase it.

@Maelkum This is part of the project workflow page no worries :D

File: content/testing.md (hunk 3 above)
Comment (Maelkum) [reply content/testing.md:68]: No no, I agree completely, I only thought that the third bullet point is a bit of an elaboration of the second one.. But all the mentioned things are definitely relevant and should be mentioned I think :)

**Review by ricardomgoncalves — APPROVED** [review]

**Pull request merged**